bundles/org.openhab.binding.wled/src/main/java/org/openhab/binding/wled/internal/WLedHandler.java

```diff
@@ -0,0 +1,456 @@
+            palleteOptions.add(new StateOption("" + counter++, value));
+        }
+        stateDescriptionProvider.setStateOptions(new ChannelUID(getThing().getUID(), CHANNEL_PALETTES), palleteOptions);
+    }
+
+    private void processState(String message) {
+        logger.trace("WLED states are:{}", message);
+        if (thing.getStatus() != ThingStatus.ONLINE) {
+            updateStatus(ThingStatus.ONLINE);
+            sendGetRequest("/json"); // fetch FX and Pallete names
+        }
+        if (message.contains("\"effects\":[")) {// JSON API reply
+            scrapeChannelOptions(message);
+            return;
+        }
+        if (message.contains("<ac>0</ac>")) {
+            updateState(CHANNEL_MASTER_CONTROLS, OnOffType.OFF);
+        } else {
+            masterBrightness = new BigDecimal(getValue(message, "<ac>", "<"));
+            updateState(CHANNEL_MASTER_CONTROLS,
+                    new PercentType(masterBrightness.divide(new BigDecimal(2.55), RoundingMode.HALF_UP)));
+        }
+        if (message.contains("<ix>0</ix>")) {
+            updateState(CHANNEL_INTENSITY, OnOffType.OFF);
+        } else {
+            BigDecimal bigTemp = new BigDecimal(getValue(message, "<ix>", "<")).divide(new BigDecimal(2.55),
+                    RoundingMode.HALF_UP);
+            updateState(CHANNEL_INTENSITY, new PercentType(bigTemp));
+        }
+        if (message.contains("<cy>1</cy>")) {
+            updateState(CHANNEL_PRESET_CYCLE, OnOffType.ON);
+        } else {
+            updateState(CHANNEL_PRESET_CYCLE, OnOffType.OFF);
+        }
+        if (message.contains("<nl>1</nl>")) {
+            updateState(CHANNEL_SLEEP, OnOffType.ON);
+        } else {
+            updateState(CHANNEL_SLEEP, OnOffType.OFF);
+        }
+        if (message.contains("<fx>")) {
+            updateState(CHANNEL_FX, new StringType(getValue(message, "<fx>", "<")));
+        }
+        if (message.contains("<sx>")) {
+            BigDecimal bigTemp = new BigDecimal(getValue(message, "<sx>", "<")).divide(new BigDecimal(2.55),
+                    RoundingMode.HALF_UP);
+            updateState(CHANNEL_SPEED, new PercentType(bigTemp));
+        }
+        if (message.contains("<fp>")) {
+            updateState(CHANNEL_PALETTES, new StringType(getValue(message, "<fp>", "<")));
+        }
+        parseColours(message);
+    }
+
+    private void sendWhite() {
+        if (hasWhite) {
+            sendGetRequest("/win&TT=1000&FX=0&CY=0&CL=hFF000000" + "&A=" + masterBrightness);
+        } else {
+            sendGetRequest("/win&TT=1000&FX=0&CY=0&CL=hFFFFFF" + "&A=" + masterBrightness);
+        }
+    }
+
+    /**
+     *
+     * @param hsb
+     * @return WLED needs the letter h followed by 2 digit HEX code for RRGGBB
+     */
+    private String createColorHex(HSBType hsb) {
+        String hex = Integer.toHexString(hsb.getRGB() & 0xffffff);
+        while (hex.length() < 6) {
+            hex = "0" + hex;
+        }
```

Comment:
When building strings in loops, `StringBuilder` should be used. In this case, you could replace the block by something like this `return String.format("h%06X", hsb.getRGB());`.

Reply:
Added your suggestion and it works, thanks much nicer.